### PR TITLE
Support creation of NoParking policy intent

### PR DESCRIPTION
**Pull request by janedotx**
## 📚 Purpose
With this PR, we're introducing a new policy-related entity: the policy intent. It's basically like a template, but somewhat more abstract. It captures the essence of the friendly, user-facing elements of a policy and could possibly have multiple implementations (different versions of the MDS spec, different ways of using policy to implement the same kind of policy--e.g., a policy that intends to ban a type of vehicle could be implemented as a count policy with a max of 0, or a speed policy with a max speed of 0). 

Basically, we're exposing only a subset of the possible fields for a policy and letting a user configure those. The remaining fields will be filled in by default settings of the particular intent. 

## 👌 Resolves:
- [ ] 🐛 JRA-834 fixes bug
- [ ] ✨ JRA-756 implements new feature

## 📦 Impacts:
*[list of packages]*

## ✅ PR Checklist
- [ ] Add or remove checklist items to suit your needs

### CODE REVIEW

File: packages/mds-policy-service/@types/intents.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+export interface BaseIntentPolicyUserFields {
+  name: string
+  description: string
+  provider_ids: Nullable<UUID[]>
+  start_date: Timestamp
+  end_date: Nullable<Timestamp>
+}
+
+export type PartialMicroMobilityStatesMap = Partial<MicroMobilityStatesToEvents>
+export type PartialMicromobilityRule = Omit<Partial<BaseRule<PartialMicroMobilityStatesMap>>, 'rule_id'>
```
Comment (avatarneil) [inline packages/mds-policy-service/@types/intents.ts:56]: Are these types necessary? Doesn't look like they're used anywhere

File: packages/mds-policy-service/@types/intents.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  no_parking: NO_PARKING_INTENT_RULE_FIXED_VALUES
+}
+
+export interface BaseIntentPolicyUserFields {
+  name: string
+  description: string
+  provider_ids: Nullable<UUID[]>
+  start_date: Timestamp
+  end_date: Nullable<Timestamp>
+}
```
[the remark this answers is not in the document]
Comment (janedotx) [reply packages/mds-policy-service/@types/intents.ts:53]: Ah right, I forgot about that type. Thanks

File: packages/mds-policy-service/@types/intents.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+export type NoParkingIntentRuleUserFields = BaseIntentRuleUserFields
+export type NoParkingIntentPolicyUserFields = BaseIntentPolicyUserFields
+
+export type NoParkingIntentDraft = IntentDraft<
+  'no_parking',
+  NoParkingIntentPolicyUserFields,
+  NoParkingIntentRuleUserFields
+>
```
[the remark this answers is not in the document]
Comment (janedotx) [reply packages/mds-policy-service/@types/intents.ts:82]: Eventually. We're not doing drafts in this current spring, but we will in the next sprint.

File: packages/mds-policy-service/@types/intents.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  rule_units: 'devices',
+  states: { available: [], non_operational: [], reserved: [], unknown: [] },
+  value_url: null,
+  vehicle_types: ['scooter', 'moped'],
+  propulsion_types: [],
+  transaction_types: null,
+  service_types: null
+}
+
+export type NoParkingIntentRuleDefaults = Omit<BaseRule<PartialMicroMobilityStatesMap, 'count'>, 'rule_id'>
```
[the remark this answers is not in the document]
Comment (janedotx) [reply packages/mds-policy-service/@types/intents.ts:41]: Throwing a const assertion on `NO_PARKING_INTENT_RULE_FIXED_VALUES` makes a lot of sense.

[the remark this answers is not in the document]
Comment (janedotx) [reply packages/mds-policy-service/@types/intents.ts:41]: You're right that this was meant to apply to any intent. I'm making the name more generic.

File: packages/mds-policy-service/@types/intents.ts (hunk 1 above)
Comment (janedotx) [reply packages/mds-policy-service/@types/intents.ts:56]: The first one is, the second is leftover from experimenting.

File: packages/mds-policy-service/@types/intents.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import { now } from '@mds-core/mds-utils'
+import type { BaseRule, MicroMobilityStatesToEvents, PolicyDomainModel } from './models'
+
+export const INTENT_TYPES = ['no_parking'] as const
```
Comment (drtyh2o) [inline packages/mds-policy-service/@types/intents.ts:4]: I tend to prefer `<const>` over `as const` these days....
```suggestion
export const INTENT_TYPES = <const>['no_parking']
```

File: packages/mds-policy-service/@types/intents.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+export type INTENT_TYPE = typeof INTENT_TYPES[number]
+
+export const BASE_POLICY_DEFAULTS = {
```
Comment (drtyh2o) [inline packages/mds-policy-service/@types/intents.ts:7]: I assume these properties align with some policy type. Would be nice to include the type rather than have this constant just be some inferred type that may or may not overlap with its intended uses.

File: packages/mds-policy-service/service/provider.ts [hunk 7]
```diff
@@ -99,5 +100,16 @@ export const PolicyServiceProvider: ServiceProvider<PolicyService, PolicyService
...
+  writePolicyIntentToPolicy: (context, intent_draft) =>
+    serviceErrorWrapper('writePolicyIntentToPolicy', async () => {
+      const policy = translateIntentToPolicy(intent_draft)
+      const { policy_id } = policy
+      await PolicyRepository.writePolicy(policy)
+      await PolicyRepository.writePolicyMetadata({
+        policy_id,
+        policy_metadata: { intent_type: intent_draft.intent_type }
+      })
+      return PolicyRepository.publishPolicy(policy.policy_id)
```
Comment (drtyh2o) [inline packages/mds-policy-service/service/provider.ts:113]: Why not use the destructured `policy_id` here?
```suggestion
      return PolicyRepository.publishPolicy(policy_id)
```

File: packages/mds-policy-service/service/provider.ts [hunk 8]
```diff
@@ -99,5 +100,16 @@ export const PolicyServiceProvider: ServiceProvider<PolicyService, PolicyService
       }
 
       return publishedPolicy
+    }),
+  writePolicyIntentToPolicy: (context, intent_draft) =>
+    serviceErrorWrapper('writePolicyIntentToPolicy', async () => {
+      const policy = translateIntentToPolicy(intent_draft)
+      const { policy_id } = policy
+      await PolicyRepository.writePolicy(policy)
```
Comment (drtyh2o) [inline packages/mds-policy-service/service/provider.ts:108]: Would this work assuming the repository returns the policy?
```suggestion
      const { policy_id } = await PolicyRepository.writePolicy(translateIntentToPolicy(intent_draft)
)
```

**Review by drtyh2o — COMMENTED** [review]
I really like where this is headed. A few comments for your consideration. Nice work.

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 9]
```diff
@@ -59,6 +59,45 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
...
+          end_date: null
+        }
+      }
+
+      const policy = await PolicyServiceClient.writePolicyIntentToPolicy(draft)
+      expect(policy).toMatchObject({
+        name: 'aname',
+        description: 'aname',
+        provider_ids: [],
+        end_date: null,
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:85]: Do we also want to assert that start_date was associated correctly?

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 10]
```diff
@@ -59,6 +59,45 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
...
+        rule_fields: {
+          geographies: [],
+          days: null,
+          start_time: null,
+          end_time: null
+        },
+        policy_fields: {
+          name: 'aname',
+          description: 'aname',
+          provider_ids: [],
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:74]: For the sake of testing it might be good to include a provider ID here and assert it, just in case there's a bug that causes provider_ids in the intent to not be correctly associated with the created policy.

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 11]
```diff
@@ -59,6 +59,45 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
...
+          start_date: now() + 50000,
+          end_date: null
+        }
+      }
+
+      const policy = await PolicyServiceClient.writePolicyIntentToPolicy(draft)
+      expect(policy).toMatchObject({
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:81]: Do we also want to assert status here? I know initially status will equal published and will eventually go to draft, but might as well assert that it's published here and not some other, unexpected status.

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 12]
```diff
@@ -59,6 +59,45 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
       })
     })
 
+    it('can translate, create, and publish a draft of a policy intent', async () => {
+      const draft: NoParkingIntentDraft = {
+        intent_type: 'no_parking',
+        rule_fields: {
+          geographies: [],
+          days: null,
+          start_time: null,
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:68]: Might want to just put a placeholder date value here and assert it just to catch any potential bugs associating start_time in the intent to the policy

File: packages/mds-policy-service/tests/policy-client.spec.ts [hunk 13]
```diff
@@ -59,6 +59,45 @@ describe('spot check unit test policy functions with SimplePolicy', () => {
       })
     })
 
+    it('can translate, create, and publish a draft of a policy intent', async () => {
+      const draft: NoParkingIntentDraft = {
+        intent_type: 'no_parking',
+        rule_fields: {
+          geographies: [],
+          days: null,
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/tests/policy-client.spec.ts:67]: May also want to provide some placeholder value for days and geographies too just to be able to assert that they were correctly assigned in the policy. Null/empty array might hide bugs that cause only default values to be passed through

File: packages/mds-policy-service/service/provider.ts [hunk 14]
```diff
@@ -99,5 +100,15 @@ export const PolicyServiceProvider: ServiceProvider<PolicyService, PolicyService
       }
 
       return publishedPolicy
+    }),
+  writePolicyIntentToPolicy: (context, intent_draft) =>
+    serviceErrorWrapper('writePolicyIntentToPolicy', async () => {
+      const { policy_id } = await PolicyRepository.writePolicy(translateIntentToPolicy(intent_draft)
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/service/provider.ts:106]: Ah I like this pattern, consolidates all of the translation logic in one place and means we don't have to create an RPC endpoint for every policy intent

File: packages/mds-policy-service/service/helpers.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  BaseIntentRuleUserFields,
+  IntentDraft,
+  INTENT_TYPE,
+  NoParkingIntentPolicyUserFields,
+  NoParkingIntentRuleUserFields
+} from '../@types/intents'
+import { BASE_POLICY_DEFAULTS, INTENT_RULE_CONSTANTS } from '../@types/intents'
+
+export function translateIntentToPolicy<
+  I extends INTENT_TYPE = 'no_parking',
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-policy-service/service/helpers.ts:14]: We may not always have a clean division between policy fields and rule fields in policy intents, so I suspect that we may run into trouble here with the type inferencing and structure of this function in the future but we shall see.

**Review by mgeorgehansen-lacuna — APPROVED** [review]
Structure of this looks great, I like the abstractions with translating policy intents to policies. I left some suggestions for increasing test coverage for some edge cases but looks solid to me otherwise.

Are we tackling GQL models in this effort? Or is that a separate task?

One point of discussion about Policy Intents: it seems like we're anticipating all policy intents to come with a separate object for policy fields and for rule fields, and that those objects should roughly match what's in MDS policy. We might end up with Policy Intents that don't match this structure, so we'll have to think about how to handle those cases once we start getting multiple different kinds of Policy Intents.

File: packages/mds-policy-service/@types/intents.ts (hunk 6 above)
Comment (janedotx) [reply packages/mds-policy-service/@types/intents.ts]: Sure thing, type added.
